piqe_ocp_lib/tests/resources/test_ocp_base.py

```diff
@@ -1,36 +1,108 @@
-    """
+
     def test_init(self, get_kubeconfig):
         base_api_obj = OcpBase(kube_config_file=get_kubeconfig)
-        assert isinstance(base_api_obj.dyn_client, DynamicClient)
         assert base_api_obj.kube_config_file is not None
-        major, minor, patch = base_api_obj.version()
-        logger.info("The obtained version is: {}.{}.{}".format(major, minor, patch))
-        assert isinstance(base_api_obj.version(), tuple)
-        assert len(base_api_obj.version()) == 3
+
+    def test_version(self, get_kubeconfig):
+        base_api_obj = OcpBase(kube_config_file=get_kubeconfig)
+        major, minor, patch = base_api_obj.version
+        logger.info("The obtained cluster version is: {}.{}.{}".format(major, minor, patch))
+        assert isinstance(base_api_obj.version, tuple)
+        assert len(base_api_obj.version) == 3
         assert isinstance(major, str) and isinstance(minor, str) and isinstance(patch, str)
 
-    def test_get_data_from_kubeconfig_4x(self, get_kubeconfig):
-        logger.info("Get data from kubeconfig file")
-        ocp_base = OcpBase(kube_config_file=get_kubeconfig)
-        kubeconfig_data = ocp_base.get_data_from_kubeconfig_v4()
-        for key, value in kubeconfig_data.items():
-            if not value:
-                assert False, f"Failed to get data for {key} from kubeconfig file"
+    def test_dynamic_client_singleton_for_ocp4x(self, get_kubeconfig):
+        logger.info("Create two instances (ocp_base1 and ocp_base2) using kubeconfig of openshift 4x cluster")
+        ocp_base1 = OcpBase(kube_config_file=get_kubeconfig)
+        logger.info("Address of ocp_base1 object is %s", ocp_base1.dyn_client)
+        ocp_base2 = OcpBase(kube_config_file=get_kubeconfig)
+        logger.info("Address of ocp_base2 object is %s", ocp_base2.dyn_client)
+
+        logger.info("Compare dynamic clients instances for kubeconfig")
+        assert ocp_base1.dyn_client is ocp_base2.dyn_client
+
+    def test_dynamic_client_singleton_multi_thread(self, get_kubeconfig):
+        def test_singleton(kube_config_file, my_queue):
+            ocp_base = OcpBase(kube_config_file=kube_config_file)
+            my_queue.put(ocp_base.dyn_client)
+
+        logger.info("Create three instances of dynamic client")
+        threads = list()
+        thread_queue = Queue()
+        kube_config_files = [get_kubeconfig, get_kubeconfig]
+        for kube_config_file in kube_config_files:
+            thread = Thread(target=test_singleton, args=(kube_config_file, thread_queue,))
+            threads.append(thread)
+            thread.start()
+
+        for thread in threads:
+            thread.join()
+
+        dynamic_client1, dynamic_client2 = thread_queue.queue
+        logger.info("Dynamic_Client1 : %s", dynamic_client1)
+        logger.info("Dynamic_Client2 : %s", dynamic_client2)
+        assert dynamic_client1 is dynamic_client2
+
+    # def test_dynamic_client_singleton_for_ocp3x(self, get_kubeconfig_3x):
```

Comment:
what is the reason that all these methods are all commented out?

Reply:
Good question...All these commented methods requires more than 1 kubeconfig file as you can see. But currently our upstream CI jenkins only support/provides 1 kubeconfig file. For that reason i have commented out these new methods. 
However, i have tested locally with two kubeconfig file and it works fine.

Reply:
I believe upstream repo is only for ocp4x plus. We/Team decided to use downstream repo for ocp3x(legacy) usecases. So you are correct, This won't work for ocp3x. We have to use downstream repo and i am not pushing this changes to downstream repo. Please correct me.

Reply:
That all makes sense, thank you for the explanation!